### PR TITLE
add magnetometer support to Android, written by @fluffrabbit

### PR DESCRIPTION
**Pull request by Marqin**
support added via Input.get_magnetometer()

I've commited this because @akien-mga closed #4010 due to rude language in commit messages.

### CODE REVIEW

**Comment by Marqin** [comment]
NOTE: I've not tested this code (and I'm not an author).

**Comment by akien-mga** [comment]
Not just due to rude language in commit messages, but to non-constructive (and yeah, overall rude) behaviour. Also the author mentioned everywhere that it was copy pasted stuff, and apparently you too haven't tested the changes, so I'm against merging it until someone really takes the time to test the code instead of just copy-pasting from StackOverflow.

**Comment by akien-mga** [comment]
Also the various ``// copypasta`` comments don't bring much. Yes, Godot's source is not much commented so adding new comments is good, but they should serve a purpose, not just "hey this was a copy paste so if it's bad, don't blame me".

**Comment by fluffrabbit** [comment]
First of all, I want to make it clear that I'm 100% in favor of testing, lest someone were to push code that didn't work. Most Android devices have magnetometers so this shouldn't be a problem for anyone who can compile Godot.

@akien-mga You're making this really personal. As I explained on IRC, which I'm pretty sure you were around to hear, I didn't copypaste a single line from SO; all the copypasta is from Godot's own existing code for accelerometers to ensure that code for the magnetometer works the exact same way. I'm going to take the high road rather than arguing with your implication that I was spending all my time bitching about the structure of Godot's deplorable codebase. That's an entirely different issue. Let's keep politics out of this.

**Comment by punto-** [comment]
This should be a pretty simple patch so there's no point in fighting over
it. We call a thing from the Android (and eventually ios) API, it gives us
a number, we give that number to the game. Even if we don't merge it right
away for whatever reason (it doesn't seem to be a popular feature), having
the pull request around means somebody who needs it can use it, because
it's simple enough that it will merge with newer changes to master or 2.0
or whatever, or at least that's what the git black magic seems to imply :)
So for now everybody is happy.

**Comment by trollworkout** [comment]
if it helps the engine it should be approved.

**Comment by Anarchid** [comment]
I would like to have this, please. I'll probably end up using it anyway, but it's cleaner if i don't have to maintain a custom local branch.

**Comment by akien-mga** [comment]
The problem is, nobody tested this code so far, and it would be meaningless to merge some code if it does not work.

See e.g. https://github.com/godotengine/godot/pull/4154/commits/2de25e28256d0fe73c7f47fcfc5f71af76205618#diff-9aa0676e56e76903d0051c7670e1ddbdR790 where the comment shows that this was not tested by the author. And indeed, the constant used does not exist in the Android API, if you google for it the only match that you get is... this pull request.

So no, this won't be merged until someone takes it seriously and works on an implementation that works (and hopefully does it for iOS too). The existing code is a useful starting point, but it's not ready for merge yet.

**Comment by Anarchid** [comment]
Hm, fair enough.

**Comment by fluffrabbit** [comment]
Well it works for me, bugs or no. It's working for somebody else too. And I'm not 100% sure you're right about the flags there, I recall there was some reason I used what I used but I've long since forgotten. I guess when you're so critical of something you're bound to find a problem sooner or later. Try reviewing the whole Godot codebase and I'm sure you'll find a few more.

**Comment by akien-mga** [comment]
> And I'm not 100% sure you're right about the flags there, I recall there was some reason I used what I used but I've long since forgotten.

http://lmgtfy.com/?q=%22ASENSOR_TYPE_MAGNETOMETER%22

> Try reviewing the whole Godot codebase and I'm sure you'll find a few more.

Which I do: https://github.com/godotengine/godot/pulls?q=is%3Apr+author%3Aakien-mga+is%3Aclosed

**Comment by Marqin** [comment]
>http://lmgtfy.com/?q=%22ASENSOR_TYPE_MAGNETOMETER%22

Try different Google, and you will find not dull critique, but a solution:
http://developer.android.com/ndk/reference/group___sensor.html

File: platform/android/godot_android.cpp [hunk 1]
```diff
@@ -789,7 +789,7 @@ void android_main(struct android_app* state) {
 	     ASENSOR_TYPE_ACCELEROMETER);
      // assuming the Android API is uniform
```
Comment (akien-mga) [inline platform/android/godot_android.cpp]: Could you also remove this misleading comment, now that we're established that the Android API is not uniform, but also removed the need for assumptions by checking the API reference?

**Comment by akien-mga** [comment]
> Try different Google, and you will find not dull critique, but a solution:
http://developer.android.com/ndk/reference/group___sensor.html

Which I found already and gave as a solution yesterday: https://github.com/godotengine/godot/pull/4154#issuecomment-216241482

So this was not dull critique, I was just responding to @fluffrabbit's suggestion that I might be doing a PITA review of this PR while I wouldn't be of others. I'm just a PITA on this PR because it's quality is lower than that of other PRs in this repo.

**Comment by fluffrabbit** [comment]
I lent a hand where it was needed, but I am by no means a career programmer. Sure, remove comments, remove trailing whitespace, remove this or that, but the point is that it works (sort of). Anyhow this is no longer my PR and I'm not going to make further PRs on this PR. @Marqin, you made this your project, but as we've seen I don't think Akien can be reasoned with. @akien-mga stop making demands. This is supposed to be a democracy, and the people have spoken.

On a slightly off topic note, since magnetometers will probably by most people be used for sensor fusion to get the device orientation, I imagine the bulk of the Godot userbase, who are (no offense) less technical than engineers, will get rather frustrated trying to hack together something that works, what with all the math and the different calibrations of different devices. I'm probably the first of them to encounter the issue of sensor fusion working one place but not another since that GDscript is my own code, so this is fair warning. Android has its own sensor fusion APIs, but of course those are different from those of iOS, and the powers that be seem to want this ported to iOS, so it's a big inter-platform rodeo. Fair warning to the devs.

**Comment by vnen** [comment]
I just noticed that my phone has a magnetic sensor so I'm gonna test this tomorrow. A quick glance can see some weird stuff, but I'm gonna review it properly tomorrow, so we can try to end this stuff once and for all (probably gonna be easier to do a new PR myself).

For @fluffrabbit and @Marqin: don't take this personally. Akien might be a bit too harsh sometimes but this is to make Godot have quality code. The second post clearly states this is not tested, so you can't blame him for rejecting the PR. And the comments are clearly not helpful and unprofessional.

**Comment by akien-mga** [comment]
> This is supposed to be a democracy, and the people have spoken.

This is not supposed to be a democracy. Open source project have always been meritocratic. The opinion of "the people" is of course very valuable and (in our project at least) highly regarded, but in the end it's the code quality that counts.

As I mentioned already, I'm all for adding magnetometer support, it sounds like a great feature that could benefit various users. But I want it done right.

**Comment by Anarchid** [comment]
> This is supposed to be a democracy, and the people have spoken.

I don't think that "the people have spoken". I was interested in this PR, but given attitudes of both authors and code quality i feel that a saner approach would be to rewrite it myself in order to avoid having unstable and incompetent maintainers.

**Comment by akien-mga** [comment]
Superseded by #4845.

**Pull request closed without merging**